launcher/src/main/java/org/apache/spark/launcher/ChildProcAppHandle.java

```diff
@@ -48,14 +48,16 @@ public synchronized void disconnect() {
 
   @Override
   public synchronized void kill() {
-    disconnect();
-    if (childProc != null) {
-      if (childProc.isAlive()) {
-        childProc.destroyForcibly();
+    if (!isDisposed()) {
+      setState(State.KILLED);
```

Comment:
so we should set the state to `KILLED` once the `kill` method is called? Even the code below fails(throw exception), the state should still be `KILLED`?

Reply:
+1，I have the same question in last review. We should figure it out.

Reply:
None of the calls below should raise exceptions. Even the socket close is wrapped in a try..catch.

Reply:
Even the order doesn't matter, I think it's more conventional to set the state at the end.

Reply:
Then the comment I made in the previous PR applies. Closing the socket / killing the child can have other implications (like changing the state) and it's easier to reason about what happens if the desired state change happens first.

Reply:
ok makes sense.